enhancements/storage/storage-hypershift.md

```diff
@@ -0,0 +1,1050 @@
+
+### Add CSI driver + csi-snapshot-controller creation directly to control-plane-operator
+
+control-plane-operator could directly run the CSI driver and csi-snapshot-controller, without cluster-storage operator,
+cluster-csi-snapshot-controller-operator or aws-ebs-csi-driver-operator.
+
+* Pro: There would be no CRs in the guest cluster, control-plane-operator could read all information it needs from
+  HosterCluster / HostedControlPlane CR.
+* Con: We would end up with two separate code bases that are able to run the operands, and we would need to update both
+  of them when fixing a bug.
+  * Esp. the CSI driver operators are complex, each of them is a bit different and we have ~9 of them (and the number is
+    growing). We would need to reimplement it in the control-plane-operator and the implementation would get out of sync
+    with the standalone OCP easily.
+
+### Rework storage operators to watch HostedCluster / HostedControlPlane
+
+All storage operators (cluster-csi-snapshot-controller-operator, cluster-storage-operator, aws-ebs-csi-driver-operator)
+may be extended to watch HostedControlPlane as their CRs and never **read** from the guest cluster, only write to it
+and eventually overwrite any user changes there.
+
+* Pro: changes in the guest cluster could not break the cluster in any way.
+* Con: Code paths for HyperShift and standalone OCP would be very different, it feels like they would be totally
+  separate operators. All bugs would need to be fixed in both places.
+  * Esp. the CSI driver operators are complex, each of them is a bit different and we have ~9 of them (and the number is
+    growing). We would need to reimplement it in the control-plane-operator and the implementation would get out of sync
+    with the standalone OCP easily.
+* Con: Or storage operators would need to be re-architected from scratch.
+
+### Rework control-plane-operator to be like CVO
+
+CVO is able to process manifests from a release payload and install them. IMO control-plane-operator could easily do the
+same instead of hardcoding the objects to create in control-plane-operator code.
+
+There would be necessary some logic that changes object namespace and e.g. generates TLS keys based on "magic"
+annotations in Service objects, similar to standalone OCP. Similarly, kubeconfigs for the guest cluster should be
+generated automatically.
+
+* Pro: very similar behavior to CVO, we would just annotate our manifests for control-plane-operator + regenerate
+  manifests for HyperShift via yaml patches. 
+* Con: significant change of control-plane-operator code.
+
+### Run everything as in standalone OCP
+
+Run cluster-csi-snapshot-controller-operator, csi-snapshot-controller, cluster-storage operator, operand
+aws-ebs-csi-driver-operator, and AWS EBS CSI driver in the guest cluster, created by CVO.
+
+It seems that Prometheus already runs in the guest cluster, so the requirement about running only user apps in the guest
+has been broken already.
```

Comment:
@csrwng how feasible is this alternative? It looks very appealing to us (storage team)!

Reply:
This alternative is less than ideal. We want to work towards the goal of reducing non-essential workloads from worker nodes. With the monitoring stack, we require at least a portion of the cmo in the data plane so we can enable things like horizontal pod autoscaling. However, the plan is to eventually make the cmo entirely optional.

Reply:
Even though this alternative is less than ideal for Hypershift, I'm afraid the main proposal could introduce bugs to existing standalone OCP clusters (see https://github.com/openshift/enhancements/pull/1202#discussion_r938055383).

In addition to that, I'll copy the [same argument used for MCO](https://github.com/openshift/enhancements/blob/0f98076064c0f2bfa92891c9079725ec677f8a17/enhancements/hypershift/node-lifecycle.md?plain=1#L117-L120): _ClusterCSIDriver_ is cluster-scoped and that doesn't fit well in the management cluster. The solution we found was to store the CR in the guest cluster, but  we're not very happy with that.

IMO, if running the _CSIControllerService_ in the management cluster is a hard requirement, we should probably skip CSO and deploy as much as we can with the control-plane-controller (just like we skip kube-controller-manager-operator, for instance). This would need to be evaluated, as I'm not sure if it's even feasible.

CC @derekwaynecarr

Reply:
Discussed outside of github. While this proposal has disadvantages pointed above, we agreed it's better to share more complex code than having separate operators and duplicate code.

There is still option to move to a separate hypershift operator later.